Polio 1601 im pipeline

quang-le opened the pull request:
Refactor Im country view to use data from json data store

Related JIRA tickets :  POLIO-1601

## Self proofreading checklist

- [x] Did I use eslint and black formatters
- [x] Is my code clear enough and well documented
- [x] Are my typescript files well typed
- [ ] New translations have been added or updated if new strings have been introduced in the frontend
- [ ] My migrations file are included
- [ ] Are there enough tests
- [ ] Documentation has been included (for new feature)

## Doc
N/A

## Changes
- Expand and rename  `LQASCountryViewset` to handle IM:
    - rename viewset, url and file to reference IM explicitly
    - enforce stricter condition on slug matching to be able to handle 3 types of IM results
    - update test and remove obsolete test
    - update LQAS API url in front-end code
- Refactor Im to follow the same structure as LQAS : overview > Map + list
    - Add `ImCountryListOverview`
    - Add `ImCountryOverview`
    - Add `ImCountryMap`
    - Move` ImSummary`
    - Delete `LqasImMap`
    - Add constant file to IM
- regroup chart related components
    - move lqas specific components out of shared folder
    - move useLqasIm in `hooks/api` folder
- Deduplicate code:
    - Add `useSelectedRound` hook to handle round selection and change
    - move `useGetCampaigns` call out of `useLqasData` and `useImData` to enable use of `useSelectedRounds`
    - Add single map component `LqasImCountryMap` and use it in `LQasOverviewContainer` and `ImOverviewContainer`
    - add `useStyle` hook in shared folder
    - Add `useMapShapes`, `useLqasImTabState` hooks
    - Add `LqasImTabs` component
- Fix some routing issues with in household IM
- Prepare code for new IM status:
    - add constant
    - add shape styling for error status (hashed pattern)
    - add status to status calculations
- Fix bug with render of round 0 in summary dropdown list
- Fix bug with `DatesIgnored` component
- Remove some useless react fragments
- Add constants for IM api urls
- improve some TS typing
- Format numbers in Summary

## How to test

- Choose a country
- Go go to polio staging. Copy the json data store entries for this country in the django admin and replace those in your DB with those from staging. Data store key formats are `im_hh_ohh_<orgunit_id>`, `im_ohh_<orgunit_id>`,`im_hh_<orgunit_id>`. Make sure to delete any duplicate data store key to avoid errors
- To make sure the data will display correctly, make sure the scopes on the local DB are the same as on staging. Same with the round/IM dates
- Once everything is setup, just go to polio> IM and test all 3 pages (Global, in household, out of household)

## Print screen / video

Out of household

https://github.com/BLSQ/iaso/assets/38907762/d6e763f8-4663-4edb-9729-6c8df559c734

In household

https://github.com/BLSQ/iaso/assets/38907762/a8fd58ca-6e07-4717-aa3c-3690d1907b25

Global

https://github.com/BLSQ/iaso/assets/38907762/dc5d9d36-97ba-43e9-a6be-4645ae96f46e


## Notes

The "Refresh" Button, that allows users to launch the pipeline and refresh data from the UI will be added as part of another PR

quang-le committed:
POLIO-1601: expand LQASCountryViewset to handle im

- rename viewset, url and file to reference IM explicitly
- enforce stricter condition on slug matching to be able to handle 3 types of IM results
- update test and remove obsolete test
- update LQAS API url in front-end code

quang-le committed:
POLIO-1601: connect IM to new datastore api

quang-le committed:
POLIO-1586: add very poor status to lqas list view

quang-le committed:
POLIO-1601: connect IM to new api endpoint

- Refactor Im to follow the same structure as LQAS : overview > Map + list
- Add ImCountryListOverview
- Add ImCountryOverview
- Add ImCountryMap
- Move ImSummary
- Delete LqasImMap
- Add constant file to IM
- Move shared components from LQAs to shared folder
- Fix bug with render of round 0 in summary dropdiown list
- Fix bug with DatesIgnored component
- Remove some useless react fragments
- Add constants for IM api urls
- improve some TS typing

quang-le committed:
POLIO-1601: add im error status

quang-le committed:
POLIO-1601: fix react unique key errors

quang-le committed:
POLIO-1601: add translations

quang-le committed:
POLIO-1601: re-organize folders

- regroup chart related components
- move lqas specific components out of shared folder
- move useLqasIm in hooks/api folder
- update imports

quang-le committed:
POLIO-1601: deduplicate code

- Add useSelectedRound hook to handle round selection and change
- move useGetCampaigns call out of useLqasData and useImData to enable use of useSelectedRounds
- Add single map component LqasImCountryMap and use it in LQasOverviewContainer and ImOverviewContainer
- add useStyle hook in shared folder
- move useImType out of index file

quang-le committed:
POLIO-1601: prepare IM for new status

- add constant
- add shape styling for error status (hashed pattern)
- add status to status calculations

quang-le committed:
POLIO-1601: more code deduplication

- Add useMapShapes, useLqasImTabState hooks
- Add LqasImTabs component
- Fix some routing issues with in household IM

quang-le committed:
POLIO-1601: fix url typo

quang-le committed:
POLIO-1601 remove obsolete test

quang-le commented on plugins/polio/tests/test_lqas_im_country.py:
```diff
@@ -96,17 +96,3 @@ def test_list_results_filtered_by_account(self):
         response_body = self.assertJSONResponse(response, 200)
         self.assertEqual(len(response_body["results"]), 1)
         self.assertEqual(response_body["results"][0]["data"], data_store_content3)
-
-    def test_list_results_filtered_by_slug(self):
-        """GET /polio/lqasmap/country/ should only show results for lqas"""
-
-        self.client.force_authenticate(self.authorized_user_regular)
-        response = self.client.get(api_url)
-        response_body = self.assertJSONResponse(response, 200)
-        self.assertEqual(len(response_body["results"]), 2)
-        self.assertEqual(response_body["results"][0]["data"], data_store_content1)
-        self.assertEqual(response_body["results"][1]["data"], data_store_content2)
-
-        self.client.force_authenticate(self.authorized_user_regular)
-        response = self.client.get(f"{api_url}{self.data_store_im.slug}/")
-        response_body = self.assertJSONResponse(response, 404)
```
The endpoint now only returns something if a `slug` is passed, so this test is obsolete

quang-le committed:
POLIO-1601: black formatting

hakifran commented on plugins/polio/js/src/constants/urls.ts:
```diff
@@ -113,15 +113,15 @@ export const polioRouteConfigs: Record<string, RouteConfig> = {
+        params: ['campaign', 'country', 'rounds', 'leftTab', 'rightTab'],
     },
-    imIhh: {
+    imHH: {
```
is there any reason you named this imHH as for other you named imIHH?

quang-le replied:
Yes, it's because I match it with what's coming from the backend to retrieve translation messages and stuff. I should actually change it everywhere

hakifran commented on plugins/polio/js/src/constants/urls.ts:
```diff
@@ -242,8 +242,8 @@ export type PolioBaseUrls = {
     imGlobal: string;
-    imIhh: string;
-    imOhh: string;
+    imHH: string;
```
Same here?

hakifran commented on plugins/polio/js/src/constants/routes.tsx:
```diff
@@ -95,14 +95,14 @@ export const imGlobalPath: RoutePath = {
 export const imIhhPath: RoutePath = {
-    baseUrl: baseUrls.imIhh,
-    routerUrl: `${baseUrls.imIhh}/*`,
+    baseUrl: baseUrls.imHH,
```
Same here the the url constant is IM_IHH but the baseUrls in named imHH, why not imIHH?

hakifran commented on plugins/polio/js/src/domains/Campaigns/MapComponent/MapComponent.js:
```diff
@@ -95,6 +95,7 @@ export const MapComponent = ({
             </Pane>
             <PaneWithPattern
                 name={`MainLayer-${name}`}
+                key={name}
```
the props for this component are :     name, patterns, patternIds and children. IMHO the key props is not need here.

quang-le replied:
You're probably right, though I had weird "unique  key prop' errors with this component before. It should be solved by giving a key to the children, but I'll leave as is for now

hakifran reviewed (approved):
LGTM apart some question and comment not blocking. @quang-le I tested the deep links and seem working fine

quang-le committed:
POLIO-1601: replace last instances of imIhh

- rename url constant
- update url path for consistency
- add redirection for old /polio/im/ihh/  path
